### ISSUE TITLE
<bit>: consider leaving only static CPU dispatch for lzcnt/bsr and tzcnt/bsr

### ISSUE THREAD

**Issue by AlexGuteniev** [microsoft/STL#2133]
Inspired by #2097 .

Try this code:
https://github.com/microsoft/STL/blob/737ce4a70d7bf04e2f230489b36a5ef21016d639/stl/inc/format#L2496-L2504

On Godbolt's compiler explorer: https://godbolt.org/z/z99vjeEhh

**Comment by AlexGuteniev** [comment]
Actually not much better. MSVC optimizer is far from perfect for this case

**Comment by StephanTLavavej** [comment]
@barcharcraz notes that these functions are typically called in perf-critical loops, so micro-optimizations can be important.

Additionally, I note that as long as we don't regress MSVC codegen quality, changes could be worthwhile if they improve Clang/LLVM codegen quality.

**Comment by AlexGuteniev** [comment]
Tried to benchmark what is better and got controversial results.

**Comment by AlexGuteniev** [comment]
The `bsr` countl_zero:
https://github.com/microsoft/STL/blob/3c2fd04d441d46ec9d914d9cbb621a3bac96c3a5/stl/inc/bit#L149-L176
Is branchy, and avoiding it seem to be a worth thing.

The `bsf` countr_zero
https://github.com/microsoft/STL/blob/3c2fd04d441d46ec9d914d9cbb621a3bac96c3a5/stl/inc/limits#L1092-L1124
Is branchless (`cmov` is used), and avoiding it seem to be not really useful.

**Comment by StephanTLavavej** [comment]
We'd like to see the specific perf data before making a decision here.

**Comment by AlexGuteniev** [comment]
So here's a program:

```
#include <bit>
#include <chrono>
#include <iostream>
#include <random>

constexpr int N = 1000;
constexpr int R = 2000000;
unsigned    int_in[N];
int         int_out;

int main() {
    std::mt19937 gen(53342);
    std::uniform_int_distribution<unsigned> int_dis(0, 1000);
    for (unsigned& element : int_in) {
        element = int_dis(gen);
    }

    auto t1 = std::chrono::steady_clock::now();
    for (int i = 0; i < R; i++) {
        for (int j = 0; j < N; j++) {
            int_out += std::countl_zero(int_in[j]);
        }
    }
    auto t2 = std::chrono::steady_clock::now();
    for (int i = 0; i < R; i++) {
        for (int j = 0; j < N; j++) {
            int_out += std::countr_zero(int_in[j]);
        }
    }
    auto t3 = std::chrono::steady_clock::now();

    auto countl = std::chrono::duration_cast<std::chrono::duration<double>>(t2 - t1);
    auto countr = std::chrono::duration_cast<std::chrono::duration<double>>(t3 - t2);

    std::cout << "countl: " << countl.count() << "\n"
        << "countr: " << countr.count() << "\n";
}
```

With `/arch:AVX2` the results are steady:
```
countl: 0.772265
countr: 0.747112
```

Currently without `/arch:AVX2` the results depend on data, as follows:
```
uniform int 0...1000:

countl: 2.04804
countr: 1.5294

uniform int 1...1000:

countl: 1.03282
countr: 1.83459
```

If I remove runtime dispatch, results depend on data as follows:
```
uniform int 0...1000

countl: 1.6128
countr: 2.04258

uniform int 1...1000:

countl: 2.67733
countr: 2.05943
```

@StephanTLavavej , I admit I have no clue what is happening here. Overall, the data suggests we should not change anything, still there's interesting case for ` 0...1000` and `countl`

It somehow overlaps with #2334 , as countl is branchy and countr is branchless.
It might suggest that we may want to force the compiler to generate branchless code (which looks possible).

I'm surprised that unexecuted branch affects the benchmark.

**Comment by AlexGuteniev** [comment]
@StephanTLavavej , I figured it out!

So overall **the advantage of lzcnt/tzcnt worth a branch with fallback, so the implementation is optimal as it is!**

All animalities I see are explained by Intel JCC Errata. Compiling with `/QIntel-jcc-erratum` fixes the performance for these cases.

I've created DevCom-1603517 to make  `/QIntel-jcc-erratum` default unless tuning for AMD.

Otherwise the data suggests that we should do dynamic dispatch, the occurrences where static-only dispatch wins are purely random. I'm closing this.
